DasharoPayloadPkg/Library/FmapLib/FmapLib.c

```diff
@@ -0,0 +1,199 @@
+/** @file
+  coreboot flash map data parsing library.
+
+Copyright (c) 2024, coreboot project<BR>
+SPDX-License-Identifier: BSD-3-Clause or GPL-2.0-only
+
+**/
+
+#include <Library/FmapLib.h>
+
+#include <Library/BaseLib.h>
+#include <Library/BaseMemoryLib.h>
+
+/* No-op on x86, but still useful to have. */
+#define le32toh(x) (x)
+#define le16toh(x) (x)
+
+STATIC
+BOOLEAN
+IsPrint (
+  IN CHAR8  Chr
+  )
+{
+  return (Chr > 32 && Chr < 126);
+}
+
+/* Make a best-effort assessment if the given fmap is real */
+STATIC
+UINTN
+is_valid_fmap (
+  IN CONST Fmap *Map
+  )
+{
+  UINTN Idx;
+
+  if (CompareMem (Map, FMAP_SIGNATURE, AsciiStrLen (FMAP_SIGNATURE)) != 0)
+    return 0;
+  /* strings containing the magic tend to fail here */
+  if (Map->ver_major != FMAP_VER_MAJOR)
+    return 0;
+  /* a basic consistency check: flash should be larger than fmap */
+  if (le32toh (Map->size) <
+    sizeof (*Map) + le16toh (Map->nareas) * sizeof (FmapArea))
+    return 0;
+
+  /* fmap-alikes along binary data tend to fail on having a valid,
+   * null-terminated string in the name field.*/
+  Idx = 0;
+  while (Idx < FMAP_STRLEN) {
+    if (Map->name[Idx] == 0)
+      break;
+    if (!IsPrint (Map->name[Idx]))
+      return 0;
+    if (Idx == FMAP_STRLEN - 1) {
+      /* name is specified to be null terminated single-word string
+       * without spaces. We did not break in the 0 test, we know it
+       * is a printable spaceless string but we're seeing FMAP_STRLEN
+       * symbols, which is one too many.
+       */
+      return 0;
+    }
+    Idx++;
+  }
+  return 1;
+
+}
+
+/* returns size of fmap data structure if successful, <0 to indicate error */
+INTN
+EFIAPI
+FmapSize (
+  IN CONST Fmap *Map
+  )
+{
+  if (!Map)
+    return -1;
+
+  return sizeof (*Map) + (le16toh (Map->nareas) * sizeof (FmapArea));
+}
+
+/* brute force linear search */
+STATIC
+INTN
+FmapLsearch (
+  IN CONST UINT8 *Image,
+  IN UINTN Len
+  )
+{
+  INTN Offset;
+  BOOLEAN FmapFound;
+
+  FmapFound = FALSE;
+
+  for (Offset = 0; Offset < Len - AsciiStrLen (FMAP_SIGNATURE); Offset++) {
```

Comment:
`is_valid_fmap` accesses fields after the signature, this would be a potential out of bounds read. Is this code copied from coreboot/cbfstool? IIRC coreboot uses defines for location and size of FMAP (which obviously isn't possible here), but cbfstool may do things differently, after all it can read any ROM, not just the one built the same layout.

```suggestion
  for (Offset = 0; Offset < Len - sizeof (Fmap); Offset++) {
```

Reply:
> `is_valid_fmap` accesses fields after the signature, this would be a potential out of bounds read.

Right.  I didn't notice it.  Other than stylistic changes, I've only done adjustments necessary to fix compilation.

> Is this code copied from coreboot/cbfstool?

Yes, https://github.com/coreboot/coreboot/blob/cb7c58f6266b10fdd1ffc3309aa929b11a81f6c5/util/cbfstool/flashmap/fmap.c#L80

https://github.com/Dasharo/edk2/compare/ab54493809413c1639ff172e219168385a576c97..a53ef1c4ebc4be7222fa234b1ce6cb61e6e4a3ee#diff-c9213a732aef6113cdcf417b51b73bea7dee0cd1ae2810f90c794db481ba9762R94